### DISCUSSION TITLE
Hub to PC write stdin not working

### DISCUSSION THREAD

**Discussion by mrinalraghupathi** [pybricks/support#1437] [Q&A] [answered]
I loaded the program below on my Spike Prime hub and then tested communicating with it over BLE using bleak. I was able to connect to the hub, start the program and stop the program. But I am not able to send data to stdin. Note I was doing this interactively from ipython and jupyterlab.

## Code on hub 

Makes the light change color based on a/b being pressed. Tested this in pybricks code and it works (i.e. if I click on the terminal area and  type a, b in the terminal the light changes color)
```python
from pybricks.hubs import InventorHub
from pybricks.parameters import Color
from pybricks.tools import wait

hub = InventorHub()

# Standard MicroPython modules
from usys import stdin, stdout
from uselect import poll

hub.display.char("A")
wait(1000)

keyboard = poll()
keyboard.register(stdin)

while True:
    while not keyboard.poll(0):
    # Optional: Do something here.
        wait(10)

    # Read one byte.
    cmd = stdin.buffer.read(1)
    if cmd == b"a":
        hub.light.on(Color.GREEN)
    elif cmd == b"b":
        hub.light.on(Color.RED)
    else:
        hub.light.off()
```

## Code to interact with hub

```python
from bleak import BleakScanner, BleakClient
from pybricksdev.ble import PYBRICKS_SERVICE_UUID
from pybricksdev.ble.pybricks import Command, PYBRICKS_COMMAND_EVENT_UUID

legohub = await BleakScanner.find_device_by_name("lovelace")
client = BleakClient(legohub)
con = await client.connect()

# This works
msg = bytearray([Command.START_USER_PROGRAM])
res = await client.write_gatt_char(PYBRICKS_COMMAND_EVENT_UUID, msg, True)

# This doesn't work (Nothing happens when the code runs. 
# The light should turn green
# note it does work from pybricks code interface through clicking on the terminal and typing input
msg = b'\x06a' # prepend the command to write to stdin
res = await client.write_gatt_char(PYBRICKS_COMMAND_EVENT_UUID, msg, True)

**Comment by dlech** [comment, marked as answer]
It could be that the "a" is being sent before the program has had a chance to actually start on the hub. I would have the hub send a "ready" signal to the host computer in the loop and have the computer wait for the ready signal before sending each command.

Example: https://github.com/orgs/pybricks/discussions/1108

**Comment by mrinalraghupathi** [comment]
Thanks! That worked. Adding a `stdout.buffer.write(b"r")` at the start of the while loop got this to work, I also added a `await client.start_notify(PYBRICKS_COMMAND_EVENT_CHAR_UUID, handle)` on the PC side which handles the signal from the hub. 

However, I then noticed that removing the hub side code `stdout.buffer.write` and keeping the `start_notify` on the PC side also made the code work. 

I'm really glad it works, but why is it that `start_notify` makes the difference vs having to actually send a signal back to the PC
.

**Reply by dlech** [reply]
I think that adding `start_notify` make it work by luck. The timing just happens to work out right. By sending a signal from the hub and waiting for that signal on the computer, then it works by design rather than by luck.

**Reply by mrinalraghupathi** [reply]
Thanks again for explaining all this in detail. Really appreciate being able to understand this better!
